File: linkdisk_flutter/副本lib/network/UdpListener.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

// 你之前已有的监听器
// abstract class DeviceFoundListener {
//   void onDeviceFound(String ip, String deviceName, String platform);
// }

class UdpListener {
  static RawDatagramSocket? _listenSocket;
  static bool _isRunning = false;
  static Timer? _broadcastTimer;

  static const int _udpPort = 54321;
  static const String _prefix = "LINKDISK_DEVICE";
  static const String _deviceIdFile = "local_device_id.txt";

  static final String _localDeviceId = _loadOrCreateDeviceId();
  static final String _localDeviceName = _getLocalDeviceName();
  static final String _localPlatform = _getPlatformName();

  // 启动监听 + 广播
  static void startListening(DeviceFoundListener listener) {
    if (_isRunning) return;
    _isRunning = true;

    _startUdpListener(listener);
    _startUdpBroadcaster();

    print("UDP监听已启动，端口：$_udpPort");
    print("本机设备ID：$_localDeviceId");
    print("本机设备名：$_localDeviceName");
    print("本机平台：$_localPlatform");
  }

  // 停止
  static void stopListening() {
    _isRunning = false;
    _listenSocket?.close();
    _broadcastTimer?.cancel();
  }

  // UDP 监听设备
  static void _startUdpListener(DeviceFoundListener listener) {
    RawDatagramSocket.bind(InternetAddress.anyIPv4, _udpPort).then((socket) {
      _listenSocket = socket;

      socket.listen((event) {
        if (!_isRunning) return;
        if (event == RawSocketEvent.read) {
          final datagram = socket.receive();
          if (datagram == null) return;

          final ip = datagram.address.address;
          final message = utf8.decode(datagram.data).trim();
          final parts = message.split('|');

          if (parts.length == 4 && parts[0] == _prefix) {
            final deviceId = parts[1];
            final deviceName = parts[2];
            final platform = parts[3];

            // 过滤自己
            if (deviceId == _localDeviceId) return;

            print("发现远程设备：$deviceName / $platform / $ip");
            listener.onDeviceFound(ip, deviceName, platform);
          }
        }
      });
    });
  }

  // 定时广播本机信息
  static void _startUdpBroadcaster() {
    RawDatagramSocket.bind(InternetAddress.anyIPv4, 0).then((socket) {
      socket.broadcastEnabled = true;
      final broadcastAddr = InternetAddress('255.255.255.255');

      _broadcastTimer = Timer.periodic(const Duration(seconds: 3), (timer) {
        if (!_isRunning) {
          timer.cancel();
          socket.close();
          return;
        }

        final msg = "$_prefix|$_localDeviceId|$_localDeviceName|$_localPlatform";
        final data = utf8.encode(msg);
        socket.send(data, broadcastAddr, _udpPort);
        print("广播已发送：$msg");
      });
    });
  }

  // 加载或创建设备唯一ID
  static String _loadOrCreateDeviceId() {
    final file = File(_deviceIdFile);
    try {
      if (file.existsSync()) {
        final id = file.readAsStringSync().trim();
        if (id.isNotEmpty) return id;
      }

      final newId = Uuid().v4();
      file.writeAsStringSync(newId);
      return newId;
    } catch (e) {
      return Uuid().v4();
    }
  }

  // 获取本机电脑名
  static String _getLocalDeviceName() {
    try {
      String? name = Platform.environment["COMPUTERNAME"];
      name ??= Platform.environment["HOSTNAME"];

      if (name == null || name.isEmpty) {
        name = InternetAddress.localHost.hostName;
      }

      name ??= "UnknownDevice";
      return name.replaceAll("|", "_");
    } catch (e) {
      return "UnknownDevice";
    }
  }

  // 获取平台名称
  static String _getPlatformName() {
    if (Platform.isWindows) return "Windows";
    if (Platform.isMacOS) return "macOS";
    if (Platform.isLinux) return "Linux";
    return Platform.operatingSystem;
  }

  // 对外提供的方法
  static String getThisDeviceName() => _localDeviceName;
  static String getThisPlatform() => _localPlatform;

  // 判断是否为本机IP
  static Future<bool> isLocalIp(String ip) async {
    try {
      for (final interface in await NetworkInterface.list()) {
        for (final addr in interface.addresses) {
          if (addr.address == ip) {
            return true;
          }
        }
      }
    } catch (e) {}
    return false;
  }
}

// 简易 UUID 实现（无需外部依赖）
class Uuid {
  String v4() {
    final rand = Random.secure();
    final bytes = Uint8List(16);
    for (int i = 0; i < 16; i++) bytes[i] = rand.nextInt(256);
    bytes[6] = (bytes[6] & 0x0F) | 0x40;
    bytes[8] = (bytes[8] & 0x3F) | 0x80;
    final hex = bytes.map((b) => b.toRadixString(16).padLeft(2, '0')).join();
    return '${hex.substring(0,8)}-${hex.substring(8,12)}-${hex.substring(12,16)}-${hex.substring(16,20)}-${hex.substring(20)}';
  }
}

File: demo/windows/Release/src/LinkDisk/network/UdpListener.java
package LinkDisk.network;

import java.io.*;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.Inet4Address;
import java.net.InetAddress;
import java.net.NetworkInterface;
import java.util.Enumeration;
import java.util.UUID;

public class UdpListener {

    private static DatagramSocket listenSocket;
    private static boolean isRunning = false;
    private static Thread listenThread;
    private static Thread broadcastThread;

    private static final int UDP_PORT = 54321;
    private static final String PREFIX = "LINKDISK_DEVICE";
    private static final String DEVICE_ID_FILE = "local_device_id.txt";

    private static final String localDeviceId = loadOrCreateDeviceId();
    private static final String localDeviceName = getLocalDeviceName();
    private static final String localPlatform = getPlatformName();

    public static void startListening(DeviceFoundListener listener) {
        if (isRunning) {
            return;
        }

        isRunning = true;

        listenThread = new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    listenSocket = new DatagramSocket(UDP_PORT);
                    byte[] buffer = new byte[1024];

                    System.out.println("UDP监听已启动，端口：" + UDP_PORT);
                    System.out.println("本机设备ID：" + localDeviceId);
                    System.out.println("本机设备名：" + localDeviceName);
                    System.out.println("本机平台：" + localPlatform);

                    while (isRunning) {
                        DatagramPacket packet = new DatagramPacket(buffer, buffer.length);

                        listenSocket.receive(packet);

                        String ip = packet.getAddress().getHostAddress();

                        String message = new String(
                                packet.getData(),
                                0,
                                packet.getLength(),
                                "UTF-8");

                        String[] parts = message.split("\\|", 4);

                        if (parts.length == 4 && PREFIX.equals(parts[0])) {

                            String deviceId = parts[1];
                            String deviceName = parts[2];
                            String platform = parts[3];

                            // 关键：过滤自己
                            if (localDeviceId.equals(deviceId)) {
                                continue;
                            }

                            System.out.println(
                                    "发现远程设备：" +
                                            deviceName + " / " +
                                            platform + " / " +
                                            ip);

                            listener.onDeviceFound(ip, deviceName, platform);
                        }
                    }

                } catch (Exception e) {
                    if (isRunning) {
                        e.printStackTrace();
                    }
                }
            }
        });

        listenThread.start();

        broadcastThread = new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    DatagramSocket broadcastSocket = new DatagramSocket();
                    broadcastSocket.setBroadcast(true);

                    System.out.println("UDP广播已启动");

                    while (isRunning) {

                        String message = PREFIX + "|" +
                                localDeviceId + "|" +
                                localDeviceName + "|" +
                                localPlatform;

                        byte[] data = message.getBytes("UTF-8");

                        DatagramPacket packet = new DatagramPacket(
                                data,
                                data.length,
                                InetAddress.getByName("255.255.255.255"),
                                UDP_PORT);

                        broadcastSocket.send(packet);

                        System.out.println("广播已发送：" + message);

                        Thread.sleep(3000);
                    }

                    broadcastSocket.close();

                } catch (Exception e) {
                    if (isRunning) {
                        e.printStackTrace();
                    }
                }
            }
        });

        broadcastThread.start();
    }

    public static void stopListening() {
        isRunning = false;

        if (listenSocket != null && !listenSocket.isClosed()) {
            listenSocket.close();
        }

        if (listenThread != null) {
            listenThread.interrupt();
        }

        if (broadcastThread != null) {
            broadcastThread.interrupt();
        }
    }

    private static String loadOrCreateDeviceId() {
        File file = new File(DEVICE_ID_FILE);

        try {
            if (file.exists()) {
                BufferedReader reader = new BufferedReader(new FileReader(file));

                String id = reader.readLine();

                reader.close();

                if (id != null && id.trim().length() > 0) {
                    return id.trim();
                }
            }

            String newId = UUID.randomUUID().toString();

            FileWriter writer = new FileWriter(file);
            writer.write(newId);
            writer.close();

            return newId;

        } catch (Exception e) {
            e.printStackTrace();
            return UUID.randomUUID().toString();
        }
    }

    private static String getLocalDeviceName() {
        try {
            String name = System.getenv("COMPUTERNAME");

            if (name == null || name.trim().length() == 0) {
                name = System.getenv("HOSTNAME");
            }

            if (name == null || name.trim().length() == 0) {
                name = InetAddress.getLocalHost().getHostName();
            }

            if (name == null || name.trim().length() == 0) {
                name = "UnknownDevice";
            }

            return name.replace("|", "_");

        } catch (Exception e) {
            return "UnknownDevice";
        }
    }

    private static String getPlatformName() {
        String os = System.getProperty("os.name").toLowerCase();

        if (os.contains("win")) {
            return "Windows";
        } else if (os.contains("mac")) {
            return "macOS";
        } else if (os.contains("linux")) {
            return "Linux";
        } else {
            return System.getProperty("os.name");
        }
    }

    public static String getThisDeviceName() {
        return localDeviceName;
    }

    public static String getThisPlatform() {
        return localPlatform;
    }

    public static boolean isLocalIp(String ip) {
        try {
            java.util.Enumeration<java.net.NetworkInterface> interfaces = java.net.NetworkInterface
                    .getNetworkInterfaces();

            while (interfaces.hasMoreElements()) {
                java.net.NetworkInterface ni = interfaces.nextElement();

                java.util.Enumeration<java.net.InetAddress> addresses = ni.getInetAddresses();

                while (addresses.hasMoreElements()) {
                    java.net.InetAddress address = addresses.nextElement();

                    if (address.getHostAddress().equals(ip)) {
                        return true;
                    }
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }

        return false;
    }

    public static String getLocalIp() {
        try {
            Enumeration<NetworkInterface> interfaces = NetworkInterface.getNetworkInterfaces();
            while (interfaces.hasMoreElements()) {
                NetworkInterface ni = interfaces.nextElement();
                if (ni.isLoopback() || !ni.isUp())
                    continue;
                Enumeration<InetAddress> addresses = ni.getInetAddresses();
                while (addresses.hasMoreElements()) {
                    InetAddress addr = addresses.nextElement();
                    if (addr instanceof Inet4Address) {
                        return addr.getHostAddress();
                    }
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }
}